lib/src/acl.rs

```diff
@@ -0,0 +1,386 @@
+use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::sync::Arc;
+
+/// Acls is a mapping of names to acl.
+#[derive(Clone, Debug, Default)]
+pub struct Acls {
+    acls: HashMap<String, Arc<Acl>>,
+}
+
+impl Acls {
+    pub fn new() -> Self {
+        Self {
+            acls: HashMap::new(),
+        }
+    }
+
+    pub fn get_acl(&self, name: &str) -> Option<&Arc<Acl>> {
+        self.acls.get(name)
+    }
+
+    pub fn insert(&mut self, name: String, acl: Acl) {
+        self.acls.insert(name, Arc::new(acl));
+    }
+}
+
+/// An acl is a collection of acl entries.
+#[derive(Debug, Default, Deserialize)]
+pub struct Acl {
+    pub(crate) entries: Vec<Entry>,
+}
+
+impl Acl {
+    /// Lookup performs a naive lookup of the given IP address
+    /// over the acls entries.
+    ///
+    /// If the IP matches multiple ACL entries, then:
+    /// - The most specific match is returned (longest mask),
```

Comment:
This looks like it gives the same results as we'd expect to see in production. Obviously the performance characteristics will be different, but I don't believe that it is a goal of Viceroy to duplicate that.